XMonad/Hooks/StatusBar.hs

```diff
@@ -358,6 +364,88 @@ statusBarPipe cmd xpp  = do
 --
 -- By using the new interface, the config becomes more declarative and there's
 -- less room for errors.
+--
+-- The only *problem* now is that the status bars will not be updated when your screen
+-- configuration changes (by plugging in a monitor, for example). Check the section
+-- on dynamic status bars for how to do that.
+
+-- $dynamic
+-- Using multiple status bars just by combining them with '<>' works well
+-- as long as the screen configuration does not change often. If it does,
+-- you should use 'dynamicSBs': by providing a function that creates
+-- status bars, it takes care of setting up the event hook, the log hook
+-- and the startup hook necessary to make the status bars, well, dynamic.
+--
+-- > xmobarTop    = statusBarPropTo "_XMONAD_LOG_1" "xmobar -x 0 ~/.config/xmobar/xmobarrc_top"    (pure ppTop)
+-- > xmobarBottom = statusBarPropTo "_XMONAD_LOG_2" "xmobar -x 0 ~/.config/xmobar/xmobarrc_bottom" (pure ppBottom)
+-- > xmobar1      = statusBarPropTo "_XMONAD_LOG_3" "xmobar -x 1 ~/.config/xmobar/xmobarrc1"       (pure pp1)
+-- >
+-- > barSpawner :: ScreenId -> IO StatusBarConfig
+-- > barSpawner 0 = pure $ xmobarTop <> xmobarBottom -- two bars on the main screen
+-- > barSpawner 1 = pure $ xmobar1
+-- > barSpawner _ = memtpy -- nothing on the rest of the screens
+-- >
+-- > main = xmonad $ dynamicSBs barSpawner (def { ... })
+--
+-- Note also that this interface can be used with one screen, or if
+-- the screen configuration doesn't change.
+
+newtype ActiveSBs = ASB {getASBs :: [(ScreenId,  StatusBarConfig)]}
+
+instance ExtensionClass ActiveSBs where
+  initialValue = ASB []
+
+-- | Given a function to create status bars, 'dynamicSBs'
+-- adds the dynamic status bar capabilities to the config.
+--
+-- Heavily inspired by "XMonad.Hooks.DynamicBars"
+dynamicSBs :: (ScreenId -> IO StatusBarConfig) -> XConfig l -> XConfig l
+dynamicSBs f conf = conf
```

Comment:
I got caught off-guard a bit by this not adding `docks` (I only read the `dynamicSB` docks, maybe it's mentioned somewhere else).

Is there a use-case where one would want to add status bars but _not_ have them recognized as such?

Reply:
Probably not. Following the same argument, should we also add `docks` (the layout modifier too? not sure) to `withSB`.

Whatever option we choose, I think we should stay consistent - this should behave like `withSB`.

Reply:
Yeah I think we should add it to `withSB` as well; bars tend to behave very weirdly if they're not recognized as bars. 

I'm kind of torn about `avoidStruts`.  The vast majority of people will probably just use it for all layouts, but there is a legitimate use-case for having certain layouts just not respect the bar.  It's probably better to leave that one to the user.

Reply:
Out of curiosity, why would one not apply `avoidStruts`?

Reply:
https://github.com/geekosaur/xmonad.hs/blob/pyanfar/xmonad.hs#L142
The quick and dirty version being to not apply it at all, but then you can't toggle it either.

Reply:
Interesting. So I guess we don't apply the layout modifier then?

Reply:
> Out of curiosity, why would one not apply `avoidStruts`?

In addition to what geekosaur already said, I can imagine someone wanting something like a `Full` layout to be completely free of struts.

I'm not sure if these use-cases are enough to justify us not applying `avoidStruts`; one may well achieve them using other means

Reply:
> Yeah I think we should add it to `withSB` as well

That's why we have `withEasySB`. There's no place for `docks` or `avoidStruts` in `withSB` or `dynamicSBs`.

Maybe later, once we have extensible/introspectable XConfig and can make combinators like `docks` and `ewmh` idempotent, then we can use them all over the place, but right now it's not safe to do that.

Reply:
Seems like people's configs will break either way; be that because an extraneous `docks` is added or because nothing is added at all.  Questions seems to be for what kind of "error case" we want to optimize.

Reply:
The "missing `docks`" case is more obvious and easier to debug, and seems to be in line with the overall xmonad philosophy of giving users a choice.

Anyway, the obvious solution here is to just make it consistent and provide both `dynamicSBs` and `dynamicEasySBs`.

Reply:
Works for me :+1:

Reply:
Sounds good 👍 I'll update it

Reply:
Alright, that should be taken care of